### PR TITLE
Sprint2

### PR DESCRIPTION
**Pull request by SanJSp**

### CODE REVIEW

File: src/lib/storage/dictionary_segment.cpp [hunk 1]
```diff
@@ -1,95 +1,157 @@
...
+    return std::make_shared<FixedWidthIntegerVector<uint32_t>>(size);
+  }
+}
+
 template <typename T>
 DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
-  // Implementation goes here
+  // retrieve value segment
+  const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(abstract_segment);
+  Assert(value_segment, "Can't encode abstract segment, because it is no value segment.");
```
Comment (postmartem) [inline src/lib/storage/dictionary_segment.cpp:28]: The abstract segment might be a value segment, just not of type T, should we reflect this in this Assert-message or assert it seperately?

File: src/lib/storage/dictionary_segment.cpp [hunk 2]
```diff
@@ -1,95 +1,157 @@
...
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _segment_nullable ? static_cast<ValueID>(0) : INVALID_VALUE_ID;
 }
 
 template <typename T>
 const T DictionarySegment<T>::value_of_value_id(const ValueID value_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  Assert(_segment_nullable && value_id != null_value_id(), "Can't retrieve value for null value.");
```
Comment (postmartem) [inline src/lib/storage/dictionary_segment.cpp]: Not sure if I am completly missing something, but shouldn't we only have null values if the segment is nullable and the value_id == null_value_id?

Comment (T4rikA) [reply src/lib/storage/dictionary_segment.cpp]: The assert should check, if a segment is nullable, that we don't want to retrieve the value for the `null_value_id()`. So it should be `!_segment_nullable && value_id != null_value_id()`

File: src/lib/storage/dictionary_segment.cpp [hunk 3]
```diff
@@ -1,95 +1,157 @@
...
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _segment_nullable ? static_cast<ValueID>(0) : INVALID_VALUE_ID;
 }
 
 template <typename T>
 const T DictionarySegment<T>::value_of_value_id(const ValueID value_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  Assert(!_segment_nullable || value_id != null_value_id(), "Can't retrieve value for null value.");
```
Comment (postmartem) [inline src/lib/storage/dictionary_segment.cpp]:
```suggestion
  Assert(!(_segment_nullable && value_id == null_value_id()), "Can't retrieve value for null value.");
```
I feel like this might be more readable/ better depict what we are trying to detect

File: src/lib/storage/chunk.hpp [hunk 4]
```diff
@@ -35,6 +35,8 @@ class Chunk : private Noncopyable {
   // Returns the segment at a given position.
   std::shared_ptr<AbstractSegment> get_segment(ColumnID column_id) const;
 
+  void add_segment_at_index(const std::shared_ptr<AbstractSegment> segment, ColumnID index);
```
Comment (kathrinthenhausen) [inline src/lib/storage/chunk.hpp:38]: Do you actually use this method anywhere? Why did you add it?

File: src/lib/storage/dictionary_segment.hpp [hunk 5]
```diff
@@ -28,7 +28,7 @@ class DictionarySegment : public AbstractSegment {
   const std::vector<T>& dictionary() const;
 
   // Returns an underlying data structure.
-  std::shared_ptr<const AbstractAttributeVector> attribute_vector() const;
+  std::shared_ptr<AbstractAttributeVector> attribute_vector() const;
```
Comment (Margarete01) [inline src/lib/storage/dictionary_segment.hpp]: Why did you change the method signature? I thought we weren't supposed to change public methods.

File: src/lib/storage/dictionary_segment.cpp [hunk 6]
```diff
@@ -1,95 +1,157 @@
...
-  Fail("Implementation is missing.");
+std::shared_ptr<AbstractAttributeVector> DictionarySegment<T>::attribute_vector() const {
+  return _attribute_vector;
 }
 
 template <typename T>
 ValueID DictionarySegment<T>::null_value_id() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _segment_nullable ? static_cast<ValueID>(0) : INVALID_VALUE_ID;
```
Comment (Margarete01) [inline src/lib/storage/dictionary_segment.cpp:103]: Interesting that you decided to return 0 as `null_value_id()` if your segment `is_nullable` instead of always returning `INVALID_VALUE_ID`.

File: src/lib/storage/dictionary_segment.cpp [hunk 7]
```diff
@@ -1,95 +1,157 @@
...
+#include <map>
+#include <set>
+#include "fixed_width_integer_vector.hpp"
+#include "type_cast.hpp"
 #include "utils/assert.hpp"
+#include "value_segment.hpp"
 
 namespace opossum {
 
+std::shared_ptr<AbstractAttributeVector> get_attribute_vector(size_t vector_size, size_t size) {
```
Comment (kathrinthenhausen) [inline src/lib/storage/dictionary_segment.cpp:12]: In my opinion it is not that nice to use hard-coded values like 8, 16, 32 (but maybe I am also wrong about that). So there are methods like `std::numeric_limits<uint8_t>::max()` which not require to convert the size into needed bits with `bid_with`.

File: src/lib/storage/fixed_width_integer_vector.cpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
+#include "fixed_width_integer_vector.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+template <typename uintX_t>
+FixedWidthIntegerVector<uintX_t>::FixedWidthIntegerVector(size_t size) : _values(size) {}
```
Comment (Margarete01) [inline src/lib/storage/fixed_width_integer_vector.cpp:7]: Nice, that you thought of initializing the vector with the needed space to reduce reallocations. The standard approach of reserving space seems to be `_values.reserve(size)`. You might want to consider that.

File: src/lib/storage/dictionary_segment.cpp [hunk 9]
```diff
@@ -1,95 +1,157 @@
...
+
 template <typename T>
 DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
-  // Implementation goes here
+  // retrieve value segment
+  const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(abstract_segment);
+  Assert(value_segment, "Can't encode abstract segment, because it is no value segment.");
+
+  auto values = value_segment->values();
+  auto value_segment_size = value_segment->size();
```
Comment (Margarete01) [inline src/lib/storage/dictionary_segment.cpp]:
```suggestion
  const auto values = value_segment->values();
  const auto value_segment_size = value_segment->size();
```

File: src/lib/storage/dictionary_segment.cpp [hunk 10]
```diff
@@ -1,95 +1,157 @@
...
+#include <set>
+#include "fixed_width_integer_vector.hpp"
+#include "type_cast.hpp"
 #include "utils/assert.hpp"
+#include "value_segment.hpp"
 
 namespace opossum {
 
+std::shared_ptr<AbstractAttributeVector> get_attribute_vector(size_t vector_size, size_t size) {
+  auto bits_needed = std::bit_width(vector_size - 1);
```
Comment (Margarete01) [inline src/lib/storage/dictionary_segment.cpp]:
```suggestion
  const auto bits_needed = std::bit_width(vector_size - 1);
```

File: src/lib/storage/dictionary_segment.cpp [hunk 11]
```diff
@@ -1,95 +1,157 @@
...
+  auto value_segment_size = value_segment->size();
+  _segment_nullable = value_segment->is_nullable();
+
+  // stores value -> id
+  auto unique_values = std::map<T, ValueID>();
+
+  auto last_index = _segment_nullable ? 1 : 0;
+  for (auto index = size_t{0}; index < value_segment_size; ++index) {
+    if (!value_segment->is_null(index)) {
+      auto value = values[index];
```
Comment (Margarete01) [inline src/lib/storage/dictionary_segment.cpp]:
```suggestion
      const auto value = values[index];
```

File: src/lib/storage/dictionary_segment.cpp [hunk 12]
```diff
@@ -1,95 +1,157 @@
...
+      if (unique_values.find(value) == unique_values.end()) {
+        unique_values[value] = last_index;
+        ++last_index;
+      }
+    }
+  }
```
Comment (kathrinthenhausen) [inline src/lib/storage/dictionary_segment.cpp]: Instead of manually building a vector with unique values you could first sort the values and then erase the duplicates like proposed here: https://stackoverflow.com/questions/1041620/whats-the-most-efficient-way-to-erase-duplicates-and-sort-a-vector

File: src/lib/storage/dictionary_segment.cpp [hunk 13]
```diff
@@ -1,95 +1,157 @@
...
+  // stores value -> id
+  auto unique_values = std::map<T, ValueID>();
+
+  auto last_index = _segment_nullable ? 1 : 0;
+  for (auto index = size_t{0}; index < value_segment_size; ++index) {
+    if (!value_segment->is_null(index)) {
+      auto value = values[index];
+      if (unique_values.find(value) == unique_values.end()) {
+        unique_values[value] = last_index;
+        ++last_index;
```
Comment (Margarete01) [inline src/lib/storage/dictionary_segment.cpp]: You could consider inserting the current value using the map insert function 
```suggestion
      const auto inserted = unique_values.insert({value, last_index});
      if (inserted.second) {++last_index;}
```

File: src/lib/storage/dictionary_segment.cpp [hunk 14]
```diff
@@ -1,95 +1,157 @@
...
+      }
+    }
+  }
+
+  for (auto [value, key] : unique_values) {
```
Comment (Margarete01) [inline src/lib/storage/dictionary_segment.cpp:48]:
```suggestion
  for (const auto [value, key] : unique_values) {
```

File: src/lib/storage/dictionary_segment.cpp [hunk 15]
```diff
@@ -1,95 +1,157 @@
...
+    _dictionary.push_back(value);
+  }
+
+  const auto attribute_vector = get_attribute_vector(unique_values.size(), value_segment_size);
+
+  for (auto index = size_t{0}; index < value_segment_size; ++index) {
+    if (value_segment->is_null(index)) {
+      attribute_vector->set(index, null_value_id());
+    } else {
+      auto dict_value = unique_values[values[index]];
```
Comment (Margarete01) [inline src/lib/storage/dictionary_segment.cpp]:
```suggestion
      const auto dict_value = unique_values[values[index]];
```

File: src/lib/storage/dictionary_segment.cpp [hunk 16]
```diff
@@ -1,95 +1,157 @@
...
+  }
+  _attribute_vector = attribute_vector;
+  return;
 }
 
 template <typename T>
 AllTypeVariant DictionarySegment<T>::operator[](const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  auto return_value = get_typed_value(chunk_offset);
```
Comment (Margarete01) [inline src/lib/storage/dictionary_segment.cpp]:
```suggestion
  const auto return_value = get_typed_value(chunk_offset);
```

File: src/lib/storage/dictionary_segment.cpp [hunk 17]
```diff
@@ -1,95 +1,157 @@
...
+  if (return_value) {
+    return *return_value;
```
Comment (kathrinthenhausen) [inline src/lib/storage/dictionary_segment.cpp:70]: Why did you decide to return a pointer(in the hpp-File it says  "Returns the value at a certain position."?

File: src/lib/storage/dictionary_segment.cpp [hunk 18]
```diff
@@ -1,95 +1,157 @@
...
+  return NULL_VALUE;
 }
 
 template <typename T>
 T DictionarySegment<T>::get(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  const auto return_value = get_typed_value(chunk_offset);
+  Assert(return_value.has_value(), "Value at position: " + std::to_string(chunk_offset) + "is NULL!");
+  return *return_value;
```
Comment (kathrinthenhausen) [inline src/lib/storage/dictionary_segment.cpp:79]: Same as above

File: src/lib/storage/table.cpp [hunk 19]
```diff
@@ -55,7 +59,7 @@ ColumnCount Table::column_count() const {
 }
 
 uint64_t Table::row_count() const {
-  return ((_chunks.size() - 1) * _target_chunk_size) + _chunks.back()->size();
+  return (chunk_count() - 1) * _chunks.front()->size() + _chunks.back()->size();
```
Comment (Margarete01) [inline src/lib/storage/table.cpp:62]: Why did you decide to write `_chunks.front()->size()` instead of `_target_chunk_size`?

File: src/lib/storage/table.cpp [hunk 20]
```diff
@@ -96,12 +100,42 @@ std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
...
+    using ColumnDataType = typename decltype(data_type_t)::type;
+    auto compressed_segment = std::make_shared<DictionarySegment<ColumnDataType>>(segment);
+    compressed_segments[segment_index] = compressed_segment;
+  });
 }
 
-// GCOVR_EXCL_STOP
+void Table::compress_chunk(const ChunkID chunk_id) {
+  auto old_chunk = get_chunk(chunk_id);
+  auto segment_count = old_chunk->column_count();
```
Comment (Margarete01) [inline src/lib/storage/table.cpp]:
```suggestion
  const auto old_chunk = get_chunk(chunk_id);
  const auto segment_count = old_chunk->column_count();
```

File: src/lib/storage/table.cpp [hunk 21]
```diff
@@ -96,12 +100,42 @@ std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
...
+  auto new_chunk = std::make_shared<Chunk>();
+  auto compressed_segments = std::vector<std::shared_ptr<AbstractSegment>>(segment_count);
+
+  auto threads = std::vector<std::thread>();
+  for (auto segment_index = ColumnID{}; segment_index < segment_count; ++segment_index) {
```
Comment (Margarete01) [inline src/lib/storage/table.cpp]: Does `ColumnID{}` result in 0? Why didn't you write `ColumnID{0}`?

File: src/lib/storage/table.cpp [hunk 22]
```diff
@@ -96,12 +100,42 @@ std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
...
+    auto type = this->column_type(segment_index);
+    auto old_segment = old_chunk->get_segment(segment_index);
```
Comment (Margarete01) [inline src/lib/storage/table.cpp]:
```suggestion
    const auto type = this->column_type(segment_index);
    const auto old_segment = old_chunk->get_segment(segment_index);
```

File: src/lib/storage/table.cpp [hunk 23]
```diff
@@ -96,12 +100,42 @@ std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
...
+    auto worker = std::thread(compress_segment, old_segment, std::ref(compressed_segments), segment_index, type);
+    threads.push_back(std::move(worker));
+  }
+  // threads join
+  for (auto thread_index = size_t{0}; thread_index < segment_count; ++thread_index) {
+    threads[thread_index].join();
+  }
```
Comment (Margarete01) [inline src/lib/storage/table.cpp]: Contributing says its better practice to use ranged loops
```suggestion
  for (auto& thread : threads) {
    threads.join();
  }
```

File: src/lib/storage/table.cpp [hunk 24]
```diff
@@ -96,12 +100,42 @@ std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
...
+
+  for (auto segment_index = ColumnID{}; segment_index < segment_count; ++segment_index) {
+    new_chunk->add_segment(compressed_segments[segment_index]);
+  }
```
Comment (Margarete01) [inline src/lib/storage/table.cpp]: Here the same as above
```suggestion
  for (const auto& segment : compressed_segments) {
    new_chunk->add_segment(segment);
  }
```

File: src/test/storage/dictionary_segment_test.cpp [hunk 25]
```diff
@@ -68,6 +68,75 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
...
+  EXPECT_EQ(dict_segment->value_of_value_id(ValueID{1}), std::string("Bill"));
+  EXPECT_EQ(dict_segment->value_of_value_id(ValueID{2}), std::string("Hasso"));
+
+  EXPECT_THROW(dict_segment->value_of_value_id(dict_segment->null_value_id()), std::logic_error);
+}
+
+TEST_F(StorageDictionarySegmentTest, MemoryUsageString) {
+  value_segment_str->append("Hello");
+  auto column_str = std::make_shared<DictionarySegment<std::string>>(value_segment_str);
+  auto dict_col_str = std::dynamic_pointer_cast<opossum::DictionarySegment<std::string>>(column_str);
```
Comment (Margarete01) [inline src/test/storage/dictionary_segment_test.cpp]:
```suggestion
  const auto column_str = std::make_shared<DictionarySegment<std::string>>(value_segment_str);
  const auto dict_col_str = std::dynamic_pointer_cast<opossum::DictionarySegment<std::string>>(column_str);
```

File: src/test/storage/dictionary_segment_test.cpp [hunk 26]
```diff
@@ -68,6 +68,75 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
...
+  for (auto index = int16_t{0}; index < UINT8_MAX + 2; ++index) {
+    value_segment_int->append(index);
+  }
+  auto column = std::make_shared<DictionarySegment<int32_t>>(value_segment_int);
+  auto dict_col = std::dynamic_pointer_cast<opossum::DictionarySegment<int32_t>>(column);
+
+  EXPECT_EQ(dict_col->estimate_memory_usage(), (UINT8_MAX + 2) * sizeof(int32_t) + (UINT8_MAX + 2) * sizeof(uint16_t));
+}
+
+TEST_F(StorageDictionarySegmentTest, MemoryUsageUInt32) {
```
Comment (Margarete01) [inline src/test/storage/dictionary_segment_test.cpp:130]: Just out of curiosity: How long does this test take you?
We considered running the same test, but it sometimes took over a minute...

File: src/test/storage/dictionary_segment_test.cpp [hunk 27]
```diff
@@ -68,6 +68,75 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
...
+
+  EXPECT_EQ(dict_col_str->estimate_memory_usage(), 1 * sizeof(std::string) + 1 * sizeof(uint8_t));
+}
+
+TEST_F(StorageDictionarySegmentTest, MemoryUsageUInt8) {
+  for (auto index = int8_t{0}; index < 100; ++index) {
+    value_segment_int->append(index);
+  }
+  auto column = std::make_shared<DictionarySegment<int32_t>>(value_segment_int);
+  auto dict_col = std::dynamic_pointer_cast<opossum::DictionarySegment<int32_t>>(column);
```
Comment (Margarete01) [inline src/test/storage/dictionary_segment_test.cpp]:
```suggestion
  const auto column = std::make_shared<DictionarySegment<int32_t>>(value_segment_int);
  const auto dict_col = std::dynamic_pointer_cast<opossum::DictionarySegment<int32_t>>(column);
```

File: src/test/storage/dictionary_segment_test.cpp [hunk 28]
```diff
@@ -68,6 +68,75 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
...
+
+  EXPECT_EQ(dict_col->estimate_memory_usage(), 100 * sizeof(int32_t) + 100 * sizeof(uint8_t));
+}
+
+TEST_F(StorageDictionarySegmentTest, MemoryUsageUInt16) {
+  for (auto index = int16_t{0}; index < UINT8_MAX + 2; ++index) {
+    value_segment_int->append(index);
+  }
+  auto column = std::make_shared<DictionarySegment<int32_t>>(value_segment_int);
+  auto dict_col = std::dynamic_pointer_cast<opossum::DictionarySegment<int32_t>>(column);
```
Comment (Margarete01) [inline src/test/storage/dictionary_segment_test.cpp]:
```suggestion
  const auto column = std::make_shared<DictionarySegment<int32_t>>(value_segment_int);
  const auto dict_col = std::dynamic_pointer_cast<opossum::DictionarySegment<int32_t>>(column);
```

**Review by Margarete01 — COMMENTED** [review]
In general your code looks nice. I like how you use a map to add values to your dictionary of the DictionarySegment.
Please keep the Contributing guidelines in mind, a lot of variables can be declared const and some for loops can be range-based.
I missed the fact, that exchanging uncompressed and compressed chunks should consider concurrent access. You should have another look into that.

File: src/lib/storage/dictionary_segment.cpp (hunk 6 above)
Comment (postmartem) [reply src/lib/storage/dictionary_segment.cpp:103]: We felt like if the segment is nullable, a NULL value is not invalid but rather just exactly that - a NULL value and thus should have a valid ID

File: src/lib/storage/dictionary_segment.hpp (hunk 5 above)
Comment (T4rikA) [reply src/lib/storage/dictionary_segment.hpp]: This was a mistake introduced by first changing the `AbstractAttributeVector `to a `vector<uint_32t>` and there we lost the const. Thanks for the catch!

File: src/lib/storage/fixed_width_integer_vector.cpp (hunk 8 above)
Comment (postmartem) [reply src/lib/storage/fixed_width_integer_vector.cpp:7]: We weren't sure what approach to choose, but since we did not find a standard approach (compare: https://stackoverflow.com/questions/8928547/vector-initialization-or-reserve), were sure that we would need this capacity and would not execute push_back e.g we found this to be the more compact and direct solution

File: src/lib/storage/table.cpp (hunk 19 above)
Comment (postmartem) [reply src/lib/storage/table.cpp:62]: This also covers the possibility of needing to compress/work with not completly full chunks

File: src/lib/storage/table.cpp (hunk 21 above)
Comment (postmartem) [reply src/lib/storage/table.cpp]: Nice catch

Comment (T4rikA) [reply src/lib/storage/table.cpp]: Nice catch!

File: src/test/storage/dictionary_segment_test.cpp (hunk 26 above)
Comment (T4rikA) [reply src/test/storage/dictionary_segment_test.cpp:130]: It takes us around 120ms

**Comment by T4rikA** [comment]
> In general your code looks nice. I like how you use a map to add values to your dictionary of the DictionarySegment. Please keep the Contributing guidelines in mind, a lot of variables can be declared const and some for loops can be range-based. I missed the fact, that exchanging uncompressed and compressed chunks should consider concurrent access. You should have another look into that.

We thought creating a new segment and then adjusting the pointer is an atomic operation, thus it shouldn't be a problem. But we will ask this during the lecture tomorrow.

**Review by kathrinthenhausen — COMMENTED** [review]
Very nice, maybe take a second look on constant variables :)
